dawn/src/dawn/CodeGen/Cuda-ico/IcoChainSizes.cpp

```diff
@@ -0,0 +1,264 @@
+  } else {
+    assert(false);
+  }
+}
+
+static std::array<grid_location_t, 3> get_cell_to_edge(const grid_location_t& cell) {
+  const auto [x, y, c] = cell;
+  if(c == 0) {
+    return {{{x, y, 0}, {x, y, 1}, {x, y, 2}}};
+  } else if(c == 1) {
+    return {{{x, y, 1}, {x + 1, y, 2}, {x, y + 1, 0}}};
+  } else {
+    assert(false);
+  }
+}
+
+int ICOChainSize(const std::vector<ast::LocationType>& chain) {
+
+  assert(1 < chain.size());
+
+  auto previous_location_type = chain[0];
+  std::unordered_set<grid_location_t> previous_locations{{0, 0, 0}};
+
+  for(chain_t::size_type i = 1; i < chain.size(); ++i) {
+
+    const auto current_location_type = chain[i];
+    std::unordered_set<grid_location_t> current_locations;
+
+    assert(previous_location_type != current_location_type);
+    const connection_t connection{previous_location_type, current_location_type};
+
+    // Vertices -> Edges
+    if(connection == connection_t{ast::LocationType::Vertices, ast::LocationType::Edges}) {
+      for(const auto& previous_location : previous_locations) {
+        const auto neighbors = get_vertex_to_edge(previous_location);
+        current_locations.insert(neighbors.cbegin(), neighbors.cend());
+      }
+      // Vertices -> Cells
+    } else if(connection == connection_t{ast::LocationType::Vertices, ast::LocationType::Cells}) {
+      for(const auto& previous_location : previous_locations) {
+        const auto neighbors = get_vertex_to_cell(previous_location);
+        current_locations.insert(neighbors.cbegin(), neighbors.cend());
+      }
+      // Edges -> Vertices
+    } else if(connection == connection_t{ast::LocationType::Edges, ast::LocationType::Vertices}) {
+      for(const auto& previous_location : previous_locations) {
+        const auto neighbors = get_edge_to_vertex(previous_location);
+        current_locations.insert(neighbors.cbegin(), neighbors.cend());
+      }
+      // Edges -> Cells
+    } else if(connection == connection_t{ast::LocationType::Edges, ast::LocationType::Cells}) {
+      for(const auto& previous_location : previous_locations) {
+        const auto neighbors = get_edge_to_cell(previous_location);
+        current_locations.insert(neighbors.cbegin(), neighbors.cend());
+      }
+      // Cells -> Vertices
+    } else if(connection == connection_t{ast::LocationType::Cells, ast::LocationType::Vertices}) {
+      for(const auto& previous_location : previous_locations) {
+        const auto neighbors = get_cell_to_vertex(previous_location);
+        current_locations.insert(neighbors.cbegin(), neighbors.cend());
+      }
+      // Cells -> Edges
+    } else if(connection == connection_t{ast::LocationType::Cells, ast::LocationType::Edges}) {
+      for(const auto& previous_location : previous_locations) {
+        const auto neighbors = get_cell_to_edge(previous_location);
+        current_locations.insert(neighbors.cbegin(), neighbors.cend());
+      }
+    } else {
+      assert(false);
+    }
```

Comment:
but what would we gain by this? wouldn't we just move the exact same boiler plate code to a different place?

Reply:
Lines from 213 to 251 would be reduced to the 4 lines above, while each function definition's header such as 
```
static std::array<grid_location_t, 6> get_vertex_to_edge(const grid_location_t& vertex)
```
would be _substituted_ by
```
map.emplace(
       connection_t{ast::LocationType::Vertices, ast::LocationType::Edges},
       [](const grid_location_t& vertex)
```

Reply:
I don't have a strong opinion on this. I think both solutions are bad for different reasons:

* static if/else is a fair amount of boilerplate code. unnecessarily verbose & repetitive
* map avoids boilerplate code, but I generally think using data-structures (like map) to get control-flow should be avoided. static if/else are simpler

Of the two, I think the first one is the _lesser bad_, since these are private functions in an implementation file.
But no strong opinion from my side.

Reply:
Ben said the current one is lesser bad so into the master it goes!